### PR TITLE
Feat/add instance segmentation

### PR DESCRIPTION
**Pull request by HonzaCuhel**
This PR includes adding the option to generate an instance-segmentation dataset.

It includes:
- OWLv2 + SlimSAM for image annotation
- Updated documentation, examples, tests
- Annotation bug fixes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -271,6 +272,23 @@ save_dir/
...
+3. Instance Segmentation Annotations (instance_segmentation_annotations.json):
+
+- Each entry corresponds to an image and contains bounding boxes, masks and labels for objects in the image.
+- Format:
+
+```bash
+{
+  "image_path": {
+    "boxes": [[x_min, y_min, x_max, y_max], ...],
+    "masks": [[x0, y0], [x1, y1], ...]
```
Comment (klemen1999) [inline README.md:284]: Is this a polyline format? Wouldn't it be one polyline per box so something like this:
```
boxes: [[x_min, y_min, x_max, y_max], ....]
masks: [[[x0, y0],[x1, y1],...], [[x0, y0],[x1, y1],...], ....]
```

File: datadreamer/utils/luxonis_dataset_converter.py [hunk 2]
```diff
@@ -17,9 +17,14 @@ class LuxonisDatasetConverter(BaseConverter):
     """Class for converting a dataset to LuxonisDataset format."""
 
     def __init__(
-        self, dataset_plugin: str = None, dataset_name: str = None, seed: int = 42
+        self,
+        dataset_plugin: str = None,
+        dataset_name: str = None,
+        seed: int = 42,
+        is_instance_segmentation: bool = False,
```
Comment (klemen1999) [inline datadreamer/utils/luxonis_dataset_converter.py:24]: If this is True then let's for now print a warning that instance segmentation will be treated as semantic segmentation until we add support for instance segmentation in luxonis-ml.

File: requirements.txt [hunk 3]
```diff
@@ -12,6 +12,6 @@ accelerate>=0.25.0
 scipy>=1.10.0
 bitsandbytes>=0.42.0
 nltk>=3.8.1
-luxonis-ml[all]>=0.3.0
+luxonis-ml[all]>=0.4.0
```
Comment (klemen1999) [inline requirements.txt:15]: Let's update to 0.4.1 as it has a hotfix.

**Review by klemen1999 — APPROVED** [review]
Added some comments, otherwise LGTM

File: datadreamer/dataset_annotation/slimsam_annotator.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            if len(boxes) == 0:
+                final_segments.append([])
+                continue
+
+            inputs = self.processor(
+                images[i], input_boxes=[boxes], return_tensors="pt"
+            ).to(self.device)
+
+            with torch.no_grad():
+                outputs = self.model(**inputs, return_dict=True)
```
Comment (sokovninn) [inline datadreamer/dataset_annotation/slimsam_annotator.py:106]: SlimSAM doesn't support batched inference?

File: README.md [hunk 5]
```diff
@@ -157,13 +157,13 @@ datadreamer --config <path-to-config>
 
 ### 🔧 Additional Parameters
 
-- `--task`: Choose between detection and classification. Default is `detection`.
+- `--task`: Choose between detection, classification, instance segmentation and semantic segmentation. Default is `detection`.
```
Comment (sokovninn) [inline README.md:160]: Should be:
 Choose between detection, classification, instance-segmentation. Default is `detection`.

File: datadreamer/pipelines/generate_dataset_from_scratch.py [hunk 6]
```diff
@@ -54,6 +58,8 @@
 
 det_annotators = {"owlv2": OWLv2Annotator}
 clf_annotators = {"clip": CLIPAnnotator}
+inst_seg_annotators = {"owlv2-slimsam": SlimSAMAnnotator}
+inst_seg_to_det = {"owlv2-slimsam": OWLv2Annotator}
```
Comment (sokovninn) [inline datadreamer/pipelines/generate_dataset_from_scratch.py:62]: Isn't the dict name confusing?

File: examples/generate_dataset_and_train_yolo.ipynb [hunk 7]
```diff
@@ -78,13 +78,13 @@
     "- `--class_names` (required): Space-separated list of object names for image generation and annotation. Example: `person moon robot`.\n",
     "- `--prompts_number` (optional): Number of prompts to generate for each object. Defaults to `10`.\n",
     "- `--annotate_only` (optional): Only annotate the images without generating new ones, prompt and image generator will be skipped. Defaults to `False`.\n",
-    "- `--task`: Choose between detection and classification. Default is `detection`.\n",
+    "- `--task`: Choose between detection, classification and instance segmentation. Default is `detection`.\n",
```
Comment (sokovninn) [inline examples/generate_dataset_and_train_yolo.ipynb:81]: Should be instance-segmentation

File: examples/generate_instance_segmentation_dataset_and_train_yolo.ipynb
Comment (sokovninn) [inline examples/generate_instance_segmentation_dataset_and_train_yolo.ipynb:1]: Let's avoid having large outputs in jupyter notebooks

File: README.md (hunk 5 above)
Comment (HonzaCuhel) [reply README.md:160]: Thanks, missed that!

File: datadreamer/pipelines/generate_dataset_from_scratch.py (hunk 6 above)
Comment (HonzaCuhel) [reply datadreamer/pipelines/generate_dataset_from_scratch.py:62]: Is `inst_seg_detectors` better, @sokovninn?

Comment (sokovninn) [reply datadreamer/pipelines/generate_dataset_from_scratch.py]: Yep, makes more sense to me

File: datadreamer/dataset_annotation/slimsam_annotator.py (hunk 4 above)
Comment (HonzaCuhel) [reply datadreamer/dataset_annotation/slimsam_annotator.py:106]: The thing is that each image can have different number of detected objects, and in that case the batched inference isn't possible straight away , so that's why I implemented it per image. But now that you've mentioned it, I thought about it again and realized that we could "padd" the bboxes with dummy bboxes, so that we can have batch inference, I'm currrently testing it. Let me know @sokovninn, if you'd find this small hack better.

File: datadreamer/dataset_annotation/owlv2_annotator.py [hunk 8]
```diff
@@ -236,6 +236,22 @@ def annotate_batch(
...
+            width = images[i].width
+            height = images[i].height
+            if width > height:
+                height_ratio = height / width
+            elif height > width:
+                width_ratio = width / height
+
+            all_boxes = [
+                box
+                / torch.tensor([width_ratio, height_ratio, width_ratio, height_ratio])
```
Comment (sokovninn) [inline datadreamer/dataset_annotation/owlv2_annotator.py:251]: For me, multiplication works here instead of division. And maybe we should move this to a separate function in the class with a short description of the transformers bug?

File: datadreamer/dataset_annotation/slimsam_annotator.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+            return SamProcessor.from_pretrained("Zigeng/SlimSAM-uniform-50")
+        return SamProcessor.from_pretrained("Zigeng/SlimSAM-uniform-77")
+
+    def annotate_batch(
+        self,
+        images: List[PIL.Image.Image],
+        boxes_batch: List[np.ndarray],
+        iou_threshold: float = 0.2,
+    ) -> List[List[List[float]]]:
+        """Annotates images for the task of instance segmentation using the FastSAM
```
Comment (sokovninn) [inline datadreamer/dataset_annotation/slimsam_annotator.py:80]: Old model name here and in a few other files

File: datadreamer/dataset_annotation/slimsam_annotator.py (hunk 4 above)
Comment (sokovninn) [reply datadreamer/dataset_annotation/slimsam_annotator.py:106]: Oh, I see. Dummy bboxes is a good solution. However, I am not sure if it will bring any boost in inference speed, but it is worth a try I think.

File: datadreamer/dataset_annotation/owlv2_annotator.py (hunk 8 above)
Comment (HonzaCuhel) [reply datadreamer/dataset_annotation/owlv2_annotator.py:251]: And does the division work for you as well or not? Because in the link with the solution the author is using division. But otherwise, yes, you're right.

Comment (sokovninn) [reply datadreamer/dataset_annotation/owlv2_annotator.py:251]: Only multiplication works for me

File: datadreamer/dataset_annotation/slimsam_annotator.py (hunk 9 above)
Comment (HonzaCuhel) [reply datadreamer/dataset_annotation/slimsam_annotator.py:80]: Thanks!

File: datadreamer/dataset_annotation/slimsam_annotator.py (hunk 4 above)
Comment (HonzaCuhel) [reply datadreamer/dataset_annotation/slimsam_annotator.py:106]: It turned out not to be faster, so not gonna use it.
